File: src/main/java/gg/mineral/practice/entity/PlayerStatus.java
package gg.mineral.practice.entity;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Getter
public enum PlayerStatus {
    FOLLOWING {
        @Override
        public boolean canFly(Profile profile) {
            return true;
        }
    }, SPECTATING {
        @Override
        public boolean canFly(Profile profile) {
            return true;
        }
    }, FIGHTING {
        @Override
        public boolean canFly(Profile profile) {
            return false;
        }
    }, KIT_EDITOR {
        @Override
        public boolean canFly(Profile profile) {
            return false;
        }
    },
    KIT_CREATOR {
        @Override
        public boolean canFly(Profile profile) {
            return false;
        }
    },
    IDLE {
        @Override
        public boolean canFly(Profile profile) {
            return profile.getPlayer().hasPermission("practice.fly");
        }
    },
    QUEUEING {
        @Override
        public boolean canFly(Profile profile) {
            return profile.getPlayer().hasPermission("practice.fly");
        }
    };

    public abstract boolean canFly(Profile profile);
}

File: src/main/kotlin/gg/mineral/practice/entity/PlayerStatus.kt
package gg.mineral.practice.entity

import org.bukkit.GameMode

enum class PlayerStatus(val gameMode: GameMode) {
    FOLLOWING(GameMode.SPECTATOR) {
        override fun canFly(profile: Profile) = true
    },
    SPECTATING(GameMode.SPECTATOR) {
        override fun canFly(profile: Profile) = true
    },
    FIGHTING(GameMode.SURVIVAL) {
        override fun canFly(profile: Profile) = false
    },
    KIT_EDITOR(GameMode.SURVIVAL) {
        override fun canFly(profile: Profile) = false
    },
    KIT_CREATOR(GameMode.CREATIVE) {
        override fun canFly(profile: Profile) = false
    },
    IDLE(GameMode.SURVIVAL) {
        override fun canFly(profile: Profile) = profile.player.hasPermission("practice.fly")
    },
    QUEUEING(GameMode.SURVIVAL) {
        override fun canFly(profile: Profile) = profile.player.hasPermission("practice.fly")
    };

    abstract fun canFly(profile: Profile): Boolean
}
